lib/salus/scanners/base.rb

```diff
@@ -112,22 +117,27 @@ def run!(salus_report:, required:, pass_on_raise:, reraise:)
         pass_on_raise ? @report.pass : @report.fail
 
         # Record the error so that the Salus report captures the issue.
-        @report.error(error_data)
-        salus_report.error(error_data)
+        record_error(error_data)
 
         raise if reraise
       ensure
-        Salus::PluginManager.send_event(:scan_executed, { salus_report: @salus_report,
-                                                          scan_report: @report })
+        @@mutex.synchronize do
+          Salus::PluginManager.send_event(:scan_executed, { salus_report: @salus_report,
+                                                            scan_report: @report })
+        end
       end
     end
 
     # Runs a command on the terminal.
-    def run_shell(command, env: {}, stdin_data: '')
+    def run_shell(command, env: {}, stdin_data: '',
+                  chdir: File.expand_path(@repository&.path_to_repo))
       # If we're passed a string, convert it to an array before passing to capture3
       command = command.split unless command.is_a?(Array)
-      Salus::PluginManager.send_event(:run_shell, command)
-      Salus::ShellResult.new(*Open3.capture3(env, *command, stdin_data: stdin_data))
+      Salus::PluginManager.send_event(:run_shell, command, chdir: chdir)
+      #  chdir: '/some/directory'
+      opts = { stdin_data: stdin_data }
+      opts[:chdir] = chdir unless chdir.nil? || chdir == "."
+      Salus::ShellResult.new(*Open3.capture3(env, *command, opts))
```

Comment:
Non-blocking question:
Is `Open3.capture3` with `opts[:chdir]` thread-safe? We know that `Dir.chdir `is not thread-safe. 🤔

Reply:
Yes, a good discussion here https://bugs.ruby-lang.org/issues/9785.  The new process spawned is uses chdir vs attempting to change the ruby process itself.